meteor/tv.py

```diff
@@ -33,53 +52,52 @@ def tv_denoise_difference_map(
-            eps=TV_STOP_TOLERANCE,
-            weight=tv_lambda,
-            max_num_iter=TV_MAX_NUM_ITER,
+        denoised_map = _tv_denoise_array(
+            map_as_array=difference_map_as_array, weight=tv_lambda
+        )
+        return negentropy(denoised_map.flatten())
+
+    optimal_lambda: float
+    if lambda_values_to_scan is not None:
+        highest_negentropy = -1e8
+        for tv_lambda in lambda_values_to_scan:
+            trial_negentropy = negentropy_objective(tv_lambda)
+            if negentropy_objective(tv_lambda) > highest_negentropy:
+                optimal_lambda = tv_lambda
+                highest_negentropy = trial_negentropy
+    else:
+        optimizer_result = minimize_scalar(
+            negentropy_objective, bracket=TV_LAMBDA_RANGE, method="golden"
         )
+        assert (
+            optimizer_result.success
+        ), "Golden minimization failed to find optimal TV lambda"
+        optimal_lambda = optimizer_result.x
 
-    optimizer_result = minimize_scalar(
-        negentropy_objective, bracket=TV_LAMBDA_RANGE, method="golden"
+    final_map = _tv_denoise_array(
+        map_as_array=difference_map_as_array, weight=optimal_lambda
     )
-    assert optimizer_result.success
-
-    optimal_lambda: float = optimizer_result.x
-
-    final_map_array: np.ndarray = denoise_tv_chambolle(
-        np.array(difference_map.grid),
-        eps=TV_STOP_TOLERANCE,
-        weight=optimal_lambda,
-        max_num_iter=TV_MAX_NUM_ITER,
+    final_map_as_ccp4 = numpy_array_to_map(
+        final_map,
+        spacegroup=difference_map_coefficients.spacegroup,
+        cell=difference_map_coefficients.cell,
     )
 
-    # TODO: we may be able to simplify the code by going directly from a numpy
-    #       array to rs.DataSet here -- right now, we go through gemmi format
-
-    ccp4_map = gm.Ccp4Map()
-
-    ccp4_map.grid = gm.FloatGrid(final_map_array)
-    ccp4_map.grid.set_unit_cell(gm.UnitCell(*difference_map_coefficients.cell))
-    ccp4_map.grid.set_size(difference_map_coefficients.get_reciprocal_grid_size())
-    ccp4_map.grid.spacegroup = gm.find_spacegroup_by_name(
-        difference_map_coefficients.space_group
-    )
-    ccp4_map.grid.symmetrize_max()
-    ccp4_map.update_ccp4_header()
-
-    high_resolution_limit = np.min(difference_map_coefficients.compute_dHKL())
-    denoised_dataset = compute_coefficients_from_map(
-        ccp4_map=ccp4_map,
-        high_resolution_limit=high_resolution_limit,
-        amplitude_label=TV_AMPLITUDE_LABEL,
-        phase_label=TV_PHASE_LABEL,
+    _, dmin = resolution_limits(difference_map_coefficients)
+    final_map_coefficients = compute_coefficients_from_map(
+        ccp4_map=final_map_as_ccp4,
+        high_resolution_limit=dmin,
+        amplitude_label=difference_map_amplitude_column,
+        phase_label=difference_map_phase_column,
     )
 
-    # ^^^ replace this with something better!
+    # ducktype for now
```

Comment:
FIXITT

Reply:
implemented an optionally-returned object that we can use to hold all the output you might want, check it out